### ISSUE TITLE
[Question] Does backtest method automatically apply inverse transformation for mappings?

### ISSUE THREAD

**Issue by guimalo** [unit8co/darts#2302]
I'm currently studying Darts and it is clear to me how the backtest method works when I pass a mapped series to it as an argument.

For example, if I write

`best_model.backtest(series[0].map(np.log1p), forecast_horizon=1, metric=mae)`

where `best_model` is a Theta model object and mae was imported from Darts, will it infer the inverse transformation (in this case, np.expm1) and apply it to the forecasted data in order to calculate the metrics or not?

Thank you in advance.

**Comment by dennisbader** [comment]
Hi @guimalo, no it does not automatically inverse transform. The method does not know that you applied a mapping to `series`.

To do this, you can first generate the historical forecasts with 

```
last_points_only = False
hfcs = best_model.historical_forecasts(
    series=series[0].map(np.log1p),
    last_points_only=last_points_only,
    ...
)
```

Then you can apply the inverse transformation to `hfc`. Depending on `last_points_only` value, you get different result for historical forecasts, and might have to iterate differntly over `hfc`:
- `last_points_only=True`: For each series (in your case 1) generates one contiguous forecast with only the last values of each historical prediction  you might have to iterate differently over the historical forecasts.
- `last_points_only=False`: For each series (in your case 1) generates a list of all possible historical forecasts each of length `forecast_horizon`.

```
hfc = [hfc.map(np.expm1) for hfc in hfcs]
```

Then you can call backtest, on the untransformed `series`, and historical forecasts `hfc`.

```
bt = best_model.backtest(
    series=series[0],
    historical_forecasts=hfc,
    last_points_only=last_points_only,
    ...
)
```

By the way, you can also create you own data transformer for this and apply it on list of target series, transform/inverse transform, etc.:

```
from sklearn.preprocessing import FunctionTransformer
from darts.dataprocessing.transformers import Scaler

trafo = FunctionTransformer(np.log1p, inverse_func = np.expm1,validate=True, check_inverse = True)
sc = Scaler(scaler=trafo)

series_tr = sc.fit_transform(series)
series_inv_tr = sc.inverse_transform(series_tr)
```

As soon as we've released the new Darts version, you'll be able to apply inverse transformation directly on the output of `historical_forecasts()`.

**Comment by guimalo** [comment]
Thanks for the clarification @dennisbader, it is very much appreciated.

If it would be possible to ask another question on this issue, I'd like to have a better understanding of how the class WindowTransformer works along with the forecasting models. Suppose I want to use the rolling mean of the 12th order lag as a feature for my model, how do I instantiate the object correctly on the code and apply it on a pipeline? When predicting with this model, will Darts handle calculations recursively in an automated manner?

Again, you and the teams help is very much appreciated.

**Comment by dennisbader** [comment]
Do you mean something like below?

You can generated the rolling mean features like this:

```
import matplotlib.pyplot as plt
import pandas as pd

from darts.datasets import AirPassengersDataset
from darts.models import LinearRegressionModel

series = AirPassengersDataset().load()

# rolling mean over 6 consecutive values
past_covariates = series.window_transform(
    transforms={
        "function": "mean",
        "mode": "rolling",
        "window": 6,
    },
)

series.plot()
past_covariates.plot()
plt.show()
```
<img width="617" alt="image" src="https://github.com/unit8co/darts/assets/84317528/b4c889d1-7936-46b2-953b-2e56a35703b6">




And then use it as `past_covariates` for model training/prediction:
```
rolling_lag = 12
output_chunk_length = 12

train, val = series[:-output_chunk_length], series[-(rolling_lag + output_chunk_length):]

# this model uses as input:
# - the last 6 months of the target series (without windowing)
# - the lagged rolling mean at lag=-12
# to directly predict the next output_chunk_length=12 values
model = LinearRegressionModel(
    lags=6,
    lags_past_covariates=[-rolling_lag],
    output_chunk_length=output_chunk_length,
)

model.fit(
    series=train,
    past_covariates=past_covariates
)

preds = model.predict(
    n=output_chunk_length,
    series=train,
    past_covariates=past_covariates,
)
val.plot()
preds.plot(label="preds")
plt.show()
```
<img width="617" alt="image" src="https://github.com/unit8co/darts/assets/84317528/8c8d1758-1617-4252-9a5d-34d0e0e04fdd">


Darts currently does not apply windowing when performing auto regression (when forecast horizon `n > output_chunk_length`). Here as well, the models do not know that you applied a window transformation.

To do this you would have to create a prediction loop that:
- generates predictions for `n=output_chunk_lengths`
- appends the predictions to the end of the input target `series`
- re-generates the windowed features from the extended `series` as new `past_covariates`
- uses the extended `series` and `past_covariates` as input for the next prediction.
- repeat all steps above until you predicted `n` points
